src/core/core/rpc_engine.cpp

```diff
@@ -1030,21 +1024,19 @@ namespace dsn {
                 }
             }
 
-            // request is forwarded, we cannot use the original rpc session
+            // request is forwarded, we cannot use the original rpc session,
+            // so use client session to send response.
             else
             {
                 dbg_dassert(response->to_address.port() > MAX_CLIENT_PORT, 
                     "target address must have named port in this case");
 
-                auto sp = task_spec::get(response->local_rpc_code);
-                auto& hdr = *response->header;
-                auto& named_nets = _client_nets[sp->rpc_call_header_format];
-                network* net = named_nets[sp->rpc_call_channel];
-
-                dassert(nullptr != net, "network not present for rpc channel '%s' with format '%s' used by rpc %s",
-                    sp->rpc_call_channel.to_string(),
-                    sp->rpc_call_header_format.to_string(),
-                    hdr.rpc_name
+                // use the header format recorded in the message
+                network* net = _client_nets[response->hdr_format][RPC_CHANNEL_TCP];
```

Comment:
Currently, for connection_oriented_network, there is only two kind of
channels:
  dassert(channel == RPC_CHANNEL_TCP || channel == RPC_CHANNEL_UDP)
So, I use constant value here to decrease overhead,

On Tue, Jun 28, 2016 at 5:35 PM, Zhenyu Guo <notifications@github.com>
wrote:

> In src/core/core/rpc_engine.cpp
> <https://github.com/imzhenyu/rDSN/pull/413#discussion_r68726256>:
>
> >              else
> >              {
> >                  dbg_dassert(response->to_address.port() > MAX_CLIENT_PORT,
> >                      "target address must have named port in this case");
> >
> > -                auto sp = task_spec::get(response->local_rpc_code);
> > -                auto& hdr = *response->header;
> > -                auto& named_nets = _client_nets[sp->rpc_call_header_format];
> > -                network* net = named_nets[sp->rpc_call_channel];
> > -
> > -                dassert(nullptr != net, "network not present for rpc channel '%s' with format '%s' used by rpc %s",
> > -                    sp->rpc_call_channel.to_string(),
> > -                    sp->rpc_call_header_format.to_string(),
> > -                    hdr.rpc_name
> > +                // use the header format recorded in the message
> > +                network* net = _client_nets[response->hdr_format][RPC_CHANNEL_TCP];
>
> why the channel becomes a constant ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/imzhenyu/rDSN/pull/413/files/06da99b9da61fcf8540ce43ae73aa020d09bc31e#r68726256>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAgQPX7TxdpLSwPdri9T-2RGwqdl5Xm-ks5qQOrigaJpZM4I8yY0>
> .
>